README.md
# The goal of this is to apply some statistics/ML to some datasets that I find interesting

# US Wind Turbines 
- [US Wind Turbine Database](https://eerscmap.usgs.gov/uswtdb/data/)
- Available as GIS, GeoJSON, & CSV 

## Findings
- The US reportedly has 58,385 wind turbines as of 2019 🤠♻️
- However, they are highly concentrated in some states, and almost nonexistent in others.
- 12 states account for 81.06% of the wind turbines in the U.S.
- 17 states account for 90.23% of the wind turbines in the U.S.
- There are 20 states that each have less than 100 wind turbines
	- Combined, they have a total of 441 wind turbines.
	- This means that combined, they account for 0.76%.. < 1% 😔
- There are 15 states that each have less than 50 wind turbines
	- Combined, they have a total of 55 wind turbines.
	- This means that combined, they account for 0.09%... 😔😔😔
![Results Graph](https://github.com/EA-Cortes/dataScience/blob/master/us_windTurbine/windTurbinesBarGraph.png)

us_windTurbine/CHANGELOG.txt
# Changelog

## [1.3.0] - 2019-01-07 (due to the partial U.S. federal government shutdown, the USGS was unable to release this update until 2/1/2019)
### Added
- Addition of 1091 turbine records. Additions include:
	- 1025 new turbines.
	- 66 existing turbines issued new case_id. Replaces old records.

### Removed
- Removed 827 turbine records. Removals include:
	- 742 decommissioned turbines.
	- 3 reclassified as "not a turbine".
	- 1 residential scale turbines (<= 50 kW) identified and removed.
	- 66 turbines issued new case_id. Replaced by new records.
	- 15 duplicate turbines removed.

### Changed
- A total of 38,536 attribute changes were made. These include:
	- 5,622 "faa_ors" updates.
	- 5,971 "faa_asn" updates.
	- 2 "usgs_pr_id" updates.
	- 710 "p_name" updates.
	- 25 "p_year" updates.
	- 3,153 "p_tnum" updates.
	- 599 "p_cap" updates.
	- 3,202 "t_manu" updates.
	- 3,683 "t_model" updates.
	- 550 "t_cap" updates.
	- 503 "t_hh" updates.
	- 551 "t_rd" updates.
	- 551 "t_rsa" updates.
	- 194 "t_ttlh" updates.
	- 746 "t_conf_atr" updates.
	- 337 "t_conf_loc" updates.
	- 6,492 "t_img_date" updates.
	- 2,882 "t_img_srce" updates.
	- 1,379 "xlong" updates.
	- 1,384 "ylat" updates.

## [1.2.0] - 2018-10-01
### Added
- Addition of 849 turbine records. Additions include:
	- 730 new turbines.
	- 119 existing turbines issued new case_id. Replaces old records.

### Removed
- Removed 310 turbine records. Removals include:
	- 128 decommissioned turbines.
	- 13 reclassified as "not a turbine".
	- 43 residential scale turbines (<= 50 kW) identified and removed
	- 119 turbines issued new case_id. Replaced by new records.
	- 7 duplicate turbines removed.

### Changed
- A total of 22,628 attribute changes were made. These include:
	- 242 "faa_ors" updates.
	- 5,924 "faa_asn" updates.
	- 5,151 "p_name" updates.
	- 1,527 "p_year" updates.
	- 3,351 "p_tnum" updates.
	- 1,120 "p_cap" updates.
	- 260 "t_manu" updates.
	- 679 "t_model" updates.
	- 372 "t_cap" updates.
	- 382 "t_hh" updates.
	- 430 "t_rd" updates.
	- 432 "t_rsa" updates.
	- 312 "t_ttlh" updates.
	- 1,429 "t_conf_atr" updates.
	- 314 "t_conf_loc" updates.
	- 495 "t_img_date" updates.
	- 102 "t_img_srce" updates.
	- 52 "xlong" updates
	- 56 "ylat" updates.

## [1.1.0] - 2018-07-05
### Added
- Addition of 1081 turbine records. Additions include:
	-551 new turbines.
	-530 added due to major retrofits (e.g., rotor / nacelle replacement). Replaces old records.

### Removed
- Removed 1071 turbine records. Removals include:
	- 477 decommissioned turbines.
	- 20 reclassified as "not a turbine".
	- 44 duplicate turbines removed.
	- 530 retired due to major retrofits. Replaced by new records.

### Changed
- A total of 41,254 attribute changes were made. These include:
	- 25,942 "xlong" updates (moved from 5 to 6 decimal places).
	- 148 "faa_ors" updates.
	- 112 "faa_asn" updates.
	- 75 "usgs_pr_id" updates.
	- 940 "p_name" updates.
	- 1,114 "p_year" updates.
	- 3,305 "p_tnum" updates.
	- 969 "p_cap" updates.
	- 400 "t_manu" updates.
	- 1,206 "t_model" updates.
	- 404 "t_cap" updates.
	- 331 "t_hh" updates.
	- 640 "t_rd" updates.
	- 636 "t_rsa" updates.
	- 851 "t_ttlh" updates.
	- 758 "t_conf_atr" updates.
	- 1,014 "t_conf_loc" updates.
	- 1,485 "t_img_date" updates.
	- 504 "t_img_srce" updates.
	- 208 "xlong" updates (unrelated to decimal precision).
	- 212 "ylat" updates.

## [1.0.0] - 2018-04-19
- First public release of USWTDB. Changelog for 1.0.0 reflects differences between USWTDB and USGS turbine database/windFarm application (2014).

### Added
- Addition of 14,693 turbine records.
- Dedicated home, data, partner, and help landing pages.
- Additional USWTDB fields include "case_id", "faa_asn", "t_fips", "t_img_date".
- Added end-user web application capabilities including:
	- Dynamic filtering and turbine statistic computation.
	- Data-driven styling.
	- Dynamic project information table.
	- Enhanced project search (with auto-complete).
	- Temporal animation controls.
	- Enhanced navigation (pitch and bearing controls).
	- Enhanced geocoder (with auto-complete).
	- Enhanced geolocation controls.
	- Extended turbine-level pop-up information.

### Changed
- Migrated legacy windFarm web application platform (Flash to JavaScript).
- Migrated geospatial service endpoints.
- USWTDB fields changed include:
	"unique_id" > "usgs_pr_id",
	"site_name" > "p_name",
	"total_turb" > "p_tnum",
	"on_year_s" > "p_year",
	"manufac" > "t_manu",
	"model" > "t_model",
	"MW_turbine" > "t_cap",
	"total_cpcy" > "p_cap",
	"total_ht" > "t_ttlh",
	"tower_h" > "t_hh",
	"rotor_dia" > "t_rd",
	"rotor_s_a" > "t_rsa",
	"usgs_ydd" > "ylat",
	"usgs_xdd" > "xlong",
	"state" > "t_state",
	"county" > "t_county",
	"conf_attr" > "t_conf_atr",
	"conf_loc" > "t_conf_loc",
	"FAA_ORS" > "faa_ors",
	"image_name" > "t_img_srce".

### Removed
- Removed 5,992 turbine records (decommissioned or reclassified as "not a turbine").
- USWTDB fields removed include "year_range", "on_year_s", "type_tower", "decommiss", "blade_l", "WENDI_name", "EIA_name", "FAA_jdate", "FAA_AGL", "image_year", "comments", "FID".


us_windTurbine/nextTask.py
import pandas as pd 

f = open("annualPurchasedByState.txt", "w")
f = open("annualPurchasedByState.txt", "a")


df = pd.read_csv("uswtdb_v1_3_20190107.csv")



# print(df.dtypes)
# print(df.head())


# df_sub = fl_df[ df['p_year'] < 2000]
# df_avgPurchaseYear =  fl_df
# print(df_avgPurchaseYear)
# print(df_sub)
# print("Formatted")

def purchasedByYear(state):
	state_df = df[ df['t_state'] == state]
	minYear = int(state_df['p_year'].min())
	maxYear = int(state_df['p_year'].max())
	output = "\nThe first wind turbine in " + state + " was purchased in: " + str(minYear) + ". The most recent was purchased in: " + str(maxYear) + ""
	f.write(output)
	purchaseList = []
	f.write("\nYear - Number of wind turbines bought")
	for year in range(minYear, maxYear+1):
		filterDF = state_df[ state_df['p_year'] == year]
		windTurbs = filterDF.size/24
		n = windTurbs
		# purchaseList.append((year, windTurbs))
		innerO = "\n" + str(year) + " - " + str(windTurbs)
		f.write(innerO)
		purchaseList.append([year, windTurbs])
	# print(purchaseList)
	return purchaseList

# 
# TXList = purchasedByYear("TX")
# CAList = purchasedByYear("CA")
# ILList = purchasedByYear("IL")


# ------ # Function that pads the list for easier output # ------ # 
def padLists(lists):
	longest = 0
	for item in lists:
		# print(len(item))
		if(len(item) > longest):
			longest = len(item)
	# print(longest)

	appYear = 0
	nn = 1
	for item in lists:
		
		if(len(item) < longest):
			
			appYear = item[0][0]
			item.reverse()
			while(len(item) < longest+1):
				item.append([appYear, 0])
				appYear -= 1
			item.reverse()

		nn += 1


def getByState(states):
	# print("Get by state)")
	
	purchasedList = []
	tempList = []
	for state in states:
		tempList.append(purchasedByYear(state))
	
	padLists(tempList)
	stateList = []
	i = 0
	for state in states:
		stateList.append([state, tempList[i]])
		i += 1
	
	
	for s in stateList:
		f.write("\n\n")
		f.write(str(s))


	return purchasedList
	

us_windTurbine/windTurbs.py
import math
import numpy as np
import scipy as sp
import pandas as pd
import matplotlib as mpl
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib.ticker import MaxNLocator
from matplotlib.ticker import FuncFormatter
import matplotlib.patches as mpatches
import matplotlib.lines as mlines
from matplotlib.lines import Line2D
from collections import namedtuple

import nextTask



f = open("output.txt", "w")
f = open("output.txt", "a")

# ----- # State Abbreviations # ----- # 
states = ["AL", "AK", "AZ", "AR", "CA", "CO", "CT", "DC", "DE", "FL", "GA", 
          "HI", "ID", "IL", "IN", "IA", "KS", "KY", "LA", "ME", "MD", 
          "MA", "MI", "MN", "MS", "MO", "MT", "NE", "NV", "NH", "NJ", 
          "NM", "NY", "NC", "ND", "OH", "OK", "OR", "PA", "RI", "SC", 
          "SD", "TN", "TX", "UT", "VT", "VA", "WA", "WV", "WI", "WY"]

# ----- # Input Data # ----- # 
df = pd.read_csv("uswtdb_v1_3_20190107.csv")


wtList = []					# List of wint Turbines per state
totalWindTurbs = 0			# Count the total wind turbines
# Filter data by states
for state in states:
	thisState = "\n" + state + " has "
	filterDF = df[ df['t_state'] == state]
	windTurbs = filterDF.size/24
	wtList.append(windTurbs)
	totalWindTurbs += windTurbs
	thisState = thisState + str(windTurbs) + " wind turbines"
	f.write(thisState)


avgWT = sum(wtList)/len(wtList)
wtCount = "There are a total of " + str(totalWindTurbs) + " wind turbines in the US!"
f.write(wtCount)

wtStdv = []
sumWT = 0 
for state in wtList:
	diff = ((state - avgWT)**2)
	wtStdv.append(diff)
	sumWT += diff
	
# print(wtList)
var = sumWT/float(len(states))
std = math.sqrt(var)	
# print(std)


# ----- # State average stats # ----- # 
outputAVG = "\nThere are an average of " + str(avgWT) + " wind turbines per state"
f.write(outputAVG)
f.write("\nHowever, only the following states are above average")

aboveAvgWT = 0
aboveAvgStates = 0
sufficientWT = 0
sufficientStates = 0
belowAvgWT = 0
belowAvgStates = 0
superLowWT = 0
superLowStates = 0
for state in states:
	thisState = "\n" + state + " has "
	filterDF = df[ df['t_state'] == state]
	windTurbs = filterDF.size/24
	
	#Threshhold
	if(windTurbs > 990):
		sufficientStates += 1
		sufficientWT += windTurbs
		thisState = thisState + str(windTurbs) + " wind turbines! "
		f.write(thisState)

	#Threshhold
		if(windTurbs > avgWT):
			aboveAvgStates += 1
			aboveAvgWT += windTurbs
			thisState = thisState + str(windTurbs) + " wind turbines"
	#Threshhold
	if(windTurbs < 100):
		if(windTurbs < 50):
			superLowStates += 1
			superLowWT += windTurbs
			thisState = thisState + str(windTurbs) + " wind turbines"
			# print(thisState)
		else:
			belowAvgStates += 1
			belowAvgWT += windTurbs
			thisState = thisState + str(windTurbs) + " wind turbines"
			# print(thisState)
	#Threshhold
	if(windTurbs < 50):
		belowAvgStates += 1
		belowAvgWT += windTurbs
		thisState = thisState + str(windTurbs) + " wind turbines"
		# print(thisState)

# ----- # Output results # ----- #
percent = 100 * (aboveAvgWT / float(totalWindTurbs))
percent = (format(percent, '.2f'))
aboveAvgOut = "\n\nThere are " + str(aboveAvgStates) + " states that have more than average wind Turbines.\nThey have a combined " + str(aboveAvgWT) + " wind turbines. \nThis means they account for " + str(percent) + "% of the wind turbines"
f.write(aboveAvgOut)

percent = 100 * (sufficientWT / float(totalWindTurbs))
percent = (format(percent, '.2f'))
aboveAvgOut = "\n\nThere are " + str(sufficientStates) + " states that have more 1000 wind Turbines.\nThey have a combined " + str(sufficientWT) + " wind turbines. \nThis means they account for " + str(percent) + "% of the wind turbines"
f.write(aboveAvgOut)

percent = 100 * (belowAvgWT / float(totalWindTurbs))
percent = (format(percent, '.2f'))
belowAvgOut = "\n\nThere are " + str(belowAvgStates) + " states have < 100 wind Turbines.\nThey have a combined " + str(belowAvgWT) + " wind turbines. \nThis means they account for " + str(percent) + "% of the wind turbines"
f.write(belowAvgOut)

percent = 100 * (superLowWT / float(totalWindTurbs))
percent = (format(percent, '.2f'))
superLowOut = "\n\nThere are " + str(superLowStates) + " states have < 50 wind Turbines.\nThey have a combined " + str(superLowWT) + " wind turbines. \nThis means they account for " + str(percent) + "% of the wind turbines"
f.write(superLowOut)

# Bar graph
n_groups = 51
x = np.arange(n_groups)
mean = []
for state in states:
	mean.append(avgWT)

def getBarColor():
	colList = []
	for wt in wtList:
		if wt > avgWT:
			colList.append('g')
		elif (wt > 900):
			colList.append('c')
		elif(wt > 300):
			colList.append('b')
		else:
			colList.append('r')
	return colList

def getN():
	colList = []
	for wt in wtList:
		if wt > avgWT:
			colList.append('Above Average')
		elif (wt > 900):
			colList.append('Slightly below average')
		elif(wt > 300):
			colList.append('Above 300')
		else:
			colList.append('Needs improvement')
	return colList

def thousands(x, pos):
    'The two args are the value and tick position'
    return '%1.0f' % (x * 1)
formatter = FuncFormatter(thousands)

fig, ax = plt.subplots(figsize=(16, 9))
ax.set_title('Wind Turbines in the US by state')
ax.set_xlabel('States')
ax.set_ylabel('# of wind turbines')
ax.yaxis.set_major_formatter(formatter)
plt.bar(x, wtList, color=getBarColor(), alpha=0.7)
plt.xticks(x, states)
plt.plot(mean, 'k--')

green_patch = mpatches.Patch(color='g', label='Above average')
cyan_patch = mpatches.Patch(color='c', label='Slightly below average')
blue_patch = mpatches.Patch(color='b', label='Below average')
red_patch = mpatches.Patch(color='r', label='Needs significant imrpovement')
white_patch =mpatches.Patch(color='w', label='Do these states even care about the planet?')
mean_line = Line2D([0], [0], color='k', lw=2, label='Average wind turbines', linestyle='--')
plt.legend(loc='upper left', handles=[green_patch, cyan_patch, blue_patch, red_patch, white_patch, mean_line])



# plt.show()

nextTask.getByState(["TX", "CA", "IL", "IA"])

# The following may be useful to analyze the data using different parameters
# ----- # Data Types # ----- #
# Case_ID
# Fea_ors
# usgs_pr_id
# t_state
# t_county
# t_fips 
# p_name
# p_year
# p_num
# p_cap
# t_manu
# t_model
# t_cap
# t_hh
# t_rd
# t_rsa
# t_ttlh
# t_conf_atr
# t_conf_loc
# t_img_date
# t_img_srce
# xlong
# ylat